From 72e412962f401f4c74839ef989a425b76a7f6274 Mon Sep 17 00:00:00 2001
From: Saahil Tomar <quic_saahtoma@quicinc.com>
Date: Tue, 27 Jun 2023 14:42:02 +0530
Subject: [PATCH] soc: qcom: Add Upperbounds check for program header

e_phnum represent the number of entries in the program header table.
So add Upperbounds check for program header and return invalid if the
number of program headers is greater than PN_XNUM (0xffff).

Change-Id: I63927e2e88a57a567a52b4eb377abe94ad3713b8
Signed-off-by: Auditya Bhattaram <quic_audityab@quicinc.com>
Signed-off-by: Saahil Tomar <quic_saahtoma@quicinc.com>
---
 drivers/soc/qcom/mdt_loader.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/soc/qcom/mdt_loader.c b/drivers/soc/qcom/mdt_loader.c
index 4d58048f3b33..d252b52b2c9f 100644
--- a/drivers/soc/qcom/mdt_loader.c
+++ b/drivers/soc/qcom/mdt_loader.c
@@ -208,7 +208,7 @@ void *qcom_mdt_read_metadata(const struct firmware *fw, size_t *data_len,
 	ehdr = (struct elf32_hdr *)fw->data;
 	phdrs = (struct elf32_phdr *)(ehdr + 1);
 
-	if (ehdr->e_phnum < 2)
+	if (ehdr->e_phnum < 2 || ehdr->e_phnum > PN_XNUM)
 		return ERR_PTR(-EINVAL);
 
 	if (phdrs[0].p_type == PT_LOAD)
-- 
2.34.1

